### PR TITLE
pkgs/misc/cups/drivers: add brother mfcl3770cdw

### PR DESCRIPTION
**Pull request by steveej**
This adds the printer driver for the Brother MFCL3770CDW.
It combines the cups-wrapper and the driver in one file which allows
sharing common variables and making it less error-prone than the other
Brother drivers in repository.

I volunteer for maintaining this one as long as I've got the model
around.

###### Motivation for this change
New driver for new hardware. 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by samueldr** [comment]
Hi!

Quick question / heads-up, depending: does your printer driver hard-code some settings, e.g. page size to A4? See #53027 and [a discourse post](https://discourse.nixos.org/t/brother-2395dw-printer-driver-setup/1790).

My Brother printer's driver, and a couple other user's is seemingly disregarding the settings set in the print. As far as I was concerned, this was hard to spot until I realized what was going on.

Thanks.

**Comment by steveeJ** [comment]
@samueldr It seems to be the case here too. I have tried to print on portrait A5 (A4 split in half) and the result was as if the paper was A4. Are you suggesting to block this PR on that issue?

**Comment by samueldr** [comment]
@steveeJ since a bunch of the drivers in Nixpkgs already share the flaw I wouldn't block.

Though it would be nice to get a taskforce (1) looking at and fixing the issue (2) making a brother driver helper/builder that handles the common formats; after all it seems that they are generally similar when close in age.

**Comment by steveeJ** [comment]
@samueldr Alright, I will take a stab at it eventually. Can't be that bad to debug a CUPS issue :upside_down_face: 

WRT this PR, CI should turn green now.

File: pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix
Comment (samueldr) [inline pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix]: Sneaky whitespace at the end of line.

File: pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix
Comment (samueldr) [inline pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix]: Another sneaky whitespace at the end of line.

File: pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix
Comment (samueldr) [inline pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix]: That's probably not ideal; see later suggestion.

File: pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix
Comment (samueldr) [inline pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix]: I think using the `$(cat $NIX_CC/nix-support/dynamic-linker)`  way to figure out the right `ld-linux.so` is probably better. Since this is called using `pkgsi686Linux.callPackage` it'll end up relying on the right ld-linux interpreter (I think). A tad annoying that (in that particular driver) they only ship 32 bit binaries.

```diff
diff --git a/pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix b/pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix
index 9fa1c052597..658fe3af77c 100644
--- a/pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix
+++ b/pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix
@@ -40,9 +40,8 @@ in rec {
         --prefix PATH : ${stdenv.lib.makeBinPath [
           coreutils ghostscript gnugrep gnused which
         ]}
-    # need to use i686 glibc here, these are 32bit proprietary binaries
-    interpreter=${pkgsi686Linux.glibc}/lib/ld-linux.so.2
-    patchelf --set-interpreter "$interpreter" $dir/lpd/brmfcl3770cdwfilter
+      patchelf --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) \
+        $dir/lpd/brmfcl3770cdwfilter
     '';
 
     meta = {
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index f43c955befc..780fe3ff2ba 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -22895,6 +22895,7 @@ in
   mfcl2740dwcupswrapper = callPackage ../misc/cups/drivers/mfcl2740dwcupswrapper { };
   mfcl2740dwlpr = callPackage ../misc/cups/drivers/mfcl2740dwlpr { };
 
+  # This driver is only available as a 32 bit proprietary binary driver
   mfcl3770cdwlpr = (pkgsi686Linux.callPackage ../misc/cups/drivers/brother/mfcl3770cdw/default.nix { }).driver;
   mfcl3770cdwcupswrapper = (pkgsi686Linux.callPackage ../misc/cups/drivers/brother/mfcl3770cdw/default.nix { }).cupswrapper;
 
```

* * *

I'm wondering what is the difference for the end-user between this way (pkgsi686Linux.callPackage) and the way the other drivers do it (callPackage + using pkgsi686Linux.glibc). Seems like only printer drivers do the whole `interpreter=${pkgsi686Linux.glibc}/lib/ld-linux.so.2` thing.

**Review by samueldr — CHANGES_REQUESTED** [review]
I don't know if the "callPackage an expression returning an attrset" is frowned upon in that kind of situation, but I don't personally see anything wrong with it.

Other than the nits and changes asked, seems good.

File: pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix
Comment (steveeJ) [reply pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix]: Who do you regard as the end-user? Looking at this after a while I would prefer to leave the details of the 32 bit inside the driver's expression

Comment (samueldr) [reply pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix]: Right, inside the expression is fine I think.

I'm thinking that hardcoding the location of ld-linux, while unlikely to change, might be an issue.

Considering the following:

```
nix-instantiate --eval -E \
    'with import <nixpkgs> {}; builtins.replaceStrings ["\n"] [""] (builtins.readFile "${pkgsi686Linux.stdenv.cc}/nix-support/dynamic-linker") == "${pkgsi686Linux.glibc}/lib/ld-linux.so.2"'
```

I don't know if there is a way to derive the `dynamic-linker` file from the i686 packages set instead of hardcoding the value. While unlikely to change, I'm not particularly fond of the hardcoded value, even though a few other derivations use the same principle. I guess that using `pkgsi686.stdenv.mkDerivation` would make `$NIX_CC` the right `cc` so the file would be available to cat as described. Though I do not either if it is bad for other reasons.

**Comment by veprbl** [comment]
I believe this driver needs to follow the same form as `mfcl2740dwlpr`/`mfcl2740dwcupswrapper` and `mfcl8690cdwlpr`/`mfcl8690cdwcupswrapper`. If we later need to do the refactoring, we do it to all drivers at once.

Having said that, I see that the mentioned drivers set interpreter path in this way:
https://github.com/NixOS/nixpkgs/blob/d484f2b7fc0834a068e8ace851faa449a03963f5/pkgs/misc/cups/drivers/mfcl8690cdwlpr/default.nix#L34
https://github.com/NixOS/nixpkgs/blob/d484f2b7fc0834a068e8ace851faa449a03963f5/pkgs/misc/cups/drivers/mfcl8690cdwlpr/default.nix#L34
So I believe that solving issue raised in https://github.com/NixOS/nixpkgs/pull/55531#discussion_r257425117 is not required to merge this PR.

I propose we do the simple refactoring and merge this.

File: pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix
Comment (steveeJ) [reply pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix]: I added the requested change so the dynamic-linker path is no longer hardcoded.

**Comment by steveeJ** [comment]
> I believe this driver needs to follow the same form as `mfcl2740dwlpr`/`mfcl2740dwcupswrapper` and `mfcl8690cdwlpr`/`mfcl8690cdwcupswrapper`. If we later need to do the refactoring, we do it to all drivers at once.

While I'm aware that this PR takes a different approach I don't agree that it needs to follow the existing one. A further refactor could take this one as an example, or unify all of them to an even different one.

**Comment by veprbl** [comment]
@steveeJ In your approach you bind `driver` attribute into your implementation of `cupswrapper` (see line 66), so one will not be able override the `cupswrapper` with a different `driver`. I don't know why you disagree with the current approach, but at least it provides a way override things using the `.override` provided by the `callPackage`.

**Comment by steveeJ** [comment]
@veprbl Making driver overrideable is a good suggestion, thanks! I think it's not a difficult change to add here. The approach is motivated by trying to minimize redundancy between the driver and the wrapper.

**Comment by steveeJ** [comment]
@samueldr please take a look at the changes I made recently, and see if your comments are addressed appropriately.

**Review by samueldr — APPROVED** [review]
As far as what I assked, it looks fine.

I assume this was tested at every step by yourself, as obviously not many users will have that particular printer :).

**Comment by steveeJ** [comment]
@samueldr 

> I assume this was tested at every step by yourself, as obviously not many users will have that particular printer :).

I've been using the printer on my system and it works OK, besides the side-effects you've mentioned in https://github.com/NixOS/nixpkgs/pull/55531#issuecomment-462160454 already. It's not bad enough to really debug the issue though.

**Comment by steveeJ** [comment]
@samueldr I don't have merge access, would you mind merging this as you've approved it? Or are you waiting for another maintainer to double-check?

File: pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix
Comment (veprbl) [inline pkgs/misc/cups/drivers/brother/mfcl3770cdw/default.nix]: Here is a way to improve overridability:
```
inherit driver;
installPhase = ''
  basedir="$driver/${reldir}"
```

**Review by veprbl — APPROVED** [review]

**Pull request merged**
